compaction/compaction_manager.cc

```diff
@@ -885,10 +885,24 @@ class compaction_manager::regular_compaction_task : public compaction_manager::t
             try {
                 bool should_update_history = this->should_update_history(descriptor.options.type());
                 sstables::compaction_result res = co_await compact_sstables(std::move(descriptor), _compaction_data, std::move(release_exhausted));
+                finish_compaction();
                 if (should_update_history) {
+                    // update_history can take a long time compared to
+                    // compaction, as a call issued on shard S1 can be
+                    // handled on shard S2. If the other shard is under
+                    // heavy load, we may unnecessarily block kicking off a
+                    // new compaction. Normally it isn't a problem, as
+                    // compactions aren't super frequent, but there were
+                    // edge cases where the described behaviour caused
+                    // compaction to fail to keep up with excessive
+                    // flushing, leading to too many sstables on disk and
+                    // OOM during a read.  There is no need to wait with
+                    // next compaction until history is updated, so release
+                    // the weight earlier to remove unnecessary
+                    // serialization.
+                    weight_r.deregister();
```

Comment:
nit: this should be outside the conditional. Otherwise, after the conditional, the state of weight registration is unspecified.

Reply:
There's almost nothing after the conditional, but one day, there may be. So better to be prepared.

Reply:
Fixed in v4